manager/middlewares/authentication.go

```diff
@@ -20,38 +19,69 @@ var AccountIDCache = struct {
 }{Values: map[string]int{}, Lock: sync.Mutex{}}
 
 // Stores or updates the account data, returning the account id
-func GetOrCreateAccount(account string) (int, error) {
+func GetOrCreateAccount(identity *utils.Identity) (int, error) {
 	rhAccount := models.RhAccount{
-		Name: account,
+		Name:  identity.GetAccountNumber(),
+		OrgID: identity.OrgID,
 	}
-	// Select, and only if not found attempt an insertion
-	err := database.Db.Where("name = ?", account).Find(&rhAccount).Error
-	if err != nil {
-		utils.Log("err", err, "name", account).Warn("Error in finding account")
+	var err error
+	if rhAccount.OrgID == "" {
+		// Identity from kafka msg, missing OrgID in msg from Inventory, find by AccountNumber
+		// https://issues.redhat.com/browse/ESSNTL-2525
+		err = database.Db.Where("name = ?", rhAccount.Name).Find(&rhAccount).Error
+		if err != nil {
+			utils.Log("err", err, "name", rhAccount.Name).Warn("Error in finding account")
+		}
+	} else {
+		// Find account by OrgID
+		err = database.Db.Where("org_id = ?", rhAccount.OrgID).Find(&rhAccount).Error
+		if err != nil {
+			utils.Log("err", err, "org_id", rhAccount.OrgID).Warn("Error in finding account")
+		}
 	}
 	if rhAccount.ID != 0 {
 		return rhAccount.ID, nil
 	}
-	err = database.OnConflictUpdate(database.Db, "name", "name").Create(&rhAccount).Error
+
+	// OrgID not found, try to find account by account_number and update org_id or create new record
+	if rhAccount.OrgID == "" {
+		// kafka msg without OrgID, create account by AccountNumber
+		err = database.OnConflictUpdate(database.Db, "name", "name").Select("name").Create(&rhAccount).Error
+		if err != nil {
+			utils.Log("err", err, "name", rhAccount.Name).Warn("Error creating account")
+		}
+		return rhAccount.ID, err
+	}
+	// create new rhAccount with OrgID
+	if rhAccount.Name == "" {
+		// avoid updating other account with name=null if AccountNumber=""
+		err = database.OnConflictUpdate(database.Db, "org_id", "org_id").Select("org_id").Create(&rhAccount).Error
```

Comment:
And similarly `.Select("org_id")` redundant here?

Reply:
likewise for Select(org_id)